pkg/trace/writer/trace.go

```diff
@@ -172,27 +175,33 @@ func (w *TraceWriter) flush() {
 		log.Errorf("Failed to serialize payload, data dropped: %v", err)
 		return
 	}
-	p := newPayload(map[string]string{
-		"Content-Type":     "application/x-protobuf",
-		"Content-Encoding": "gzip",
-		headerLanguages:    strings.Join(info.Languages(), "|"),
-	})
-	gzipw, err := gzip.NewWriterLevel(p.body, gzip.BestSpeed)
-	if err != nil {
-		// it will never happen, unless an invalid compression is chosen;
-		// we know gzip.BestSpeed is valid.
-		log.Errorf("gzip.NewWriterLevel: %d", err)
-		return
-	}
-	gzipw.Write(b)
-	gzipw.Close()
 
 	atomic.AddInt64(&w.stats.BytesUncompressed, int64(len(b)))
 	atomic.AddInt64(&w.stats.BytesEstimated, int64(w.bufferedSize))
 
-	for _, sender := range w.senders {
-		sender.Push(p)
-	}
+	w.wg.Add(1)
+	go func() {
+		defer timing.Since("datadog.trace_agent.trace_writer.compress_ms", time.Now())
+		defer w.wg.Done()
+		p := newPayload(map[string]string{
```

Comment:
It'll probably also help in terms of blocking time to parallelize `proto.Marshal` (although it might hurt allocations since doing so would mean we'd need to use new slices when we `defer w.resetBuffer()` (or copy `w.traces` and `w.events` before doing the marshalling).

Reply:
Maybe. I thought about it, but it would complicate things a bit. I think we should take it one at a time. If this turns out to be a bottleneck again, we can investigate how we can parallelize the proto marshalling too.